folding-schemes/src/folding/circuits/cyclefold.rs

```diff
@@ -365,15 +372,96 @@ where
     }
 }
 
+/// Folds the given cyclefold circuit and its instances. This method is isolated from any folding
+/// scheme struct because it is used both by Nova & HyperNova's CycleFold.
+#[allow(clippy::type_complexity)]
+pub fn fold_cyclefold_circuit<C1, GC1, C2, GC2, FC, CS1, CS2>(
+    poseidon_config: &PoseidonConfig<C1::ScalarField>,
+    cf_r1cs: R1CS<C2::ScalarField>,
+    cf_cs_params: CS2::ProverParams,
+    cf_W_i: Witness<C2>,           // witness of the running instance
+    cf_U_i: CommittedInstance<C2>, // running instance
+    cf_u_i_x: Vec<C2::ScalarField>,
+    cf_circuit: CycleFoldCircuit<C1, GC1>,
+) -> Result<
+    (
+        Witness<C2>,
+        CommittedInstance<C2>, // u_i
+        Witness<C2>,           // W_i1
+        CommittedInstance<C2>, // U_i1
+        C2,                    // cmT
+        C2::ScalarField,       // r_Fq
+    ),
+    Error,
+>
+where
+    C1: CurveGroup,
+    GC1: CurveVar<C1, CF2<C1>> + ToConstraintFieldGadget<CF2<C1>>,
+    C2: CurveGroup,
+    GC2: CurveVar<C2, CF2<C2>> + ToConstraintFieldGadget<CF2<C2>>,
+    FC: FCircuit<C1::ScalarField>,
+    CS1: CommitmentScheme<C1>,
+    CS2: CommitmentScheme<C2>,
+    <C1 as CurveGroup>::BaseField: PrimeField,
+    <C2 as CurveGroup>::BaseField: PrimeField,
+    <C1 as Group>::ScalarField: Absorb,
+    <C2 as Group>::ScalarField: Absorb,
+    C1: CurveGroup<BaseField = C2::ScalarField, ScalarField = C2::BaseField>,
+    for<'a> &'a GC1: GroupOpsBounds<'a, C1, GC1>,
+    for<'a> &'a GC2: GroupOpsBounds<'a, C2, GC2>,
+{
+    let cs2 = ConstraintSystem::<C1::BaseField>::new_ref();
+    cf_circuit.generate_constraints(cs2.clone())?;
+
+    let cs2 = cs2.into_inner().ok_or(Error::NoInnerConstraintSystem)?;
+    let (cf_w_i, cf_x_i) = extract_w_x::<C1::BaseField>(&cs2);
+    if cf_x_i != cf_u_i_x {
+        return Err(Error::NotEqual);
+    }
+
+    #[cfg(test)]
+    if cf_x_i.len() != CF_IO_LEN {
+        return Err(Error::NotExpectedLength(cf_x_i.len(), CF_IO_LEN));
+    }
+
+    // fold cyclefold instances
+    let cf_w_i = Witness::<C2>::new(cf_w_i.clone(), cf_r1cs.A.n_rows);
+    let cf_u_i: CommittedInstance<C2> = cf_w_i.commit::<CS2>(&cf_cs_params, cf_x_i.clone())?;
+
+    // compute T* and cmT* for CycleFoldCircuit
+    let (cf_T, cf_cmT) = NIFS::<C2, CS2>::compute_cyclefold_cmT(
+        &cf_cs_params,
+        &cf_r1cs,
+        &cf_w_i,
+        &cf_u_i,
+        &cf_W_i,
+        &cf_U_i,
+    )?;
+
+    let cf_r_bits = CycleFoldChallengeGadget::<C2, GC2>::get_challenge_native(
+        poseidon_config,
+        cf_U_i.clone(),
+        cf_u_i.clone(),
+        cf_cmT,
+    )?;
+    let cf_r_Fq = C1::BaseField::from_bigint(BigInteger::from_bits_le(&cf_r_bits))
+        .ok_or(Error::OutOfBounds)?;
```

Comment:
agree that it should never happen, but prefer to not use `unwrap` and return a proper error, to avoid the code panicking in a deployed server, no?
Although sometimes I think that it would be better to just panic in a server so the user notices that is not working, but on the same time my thought is that better to not break the deployment and report the error, and the user of the library can choose if they want it to panic or not.

Reply:
In that case, I'd go for `expect` which allows to add some extra explanation for the error.

Reply:
but, wouldn't `expect` panic (with a message) as the `unwrap`, breaking the runtime of the server? I would prefer that the user of the lib can choose if they want the code to panic or to use the returned error.

Reply:
That's the point no? If the error we think is irrecoverable, why leaving a choice to do it wrong?

Reply:
right! I agree, updated